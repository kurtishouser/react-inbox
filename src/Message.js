import React, {Component} from 'react';
import Label from './Label.js';

class Message extends Component {

  constructor(props) {
    super(props);

    this.toggleSelected = this.toggleSelected.bind(this);
    this.toggleStarred = this.toggleStarred.bind(this);
  }

  toggleSelected () {
    let messageId = this.props.id;
    let status = !this.props.selected;
    this.props.updateSelectedStatus(messageId, status);
    // let message = this.props;
    // this.props.toggleProperty(message, 'selected');
  }

  toggleStarred() {
    let messageId = this.props.id;
    let status = !this.props.starred;
    this.props.updateStarredStatus(messageId, status);
    // let message = this.props;
    // this.props.toggleProperty(message, 'starred');
  }

  render() {
    const read = this.props.read ? 'read' : 'unread';
    const selected = this.props.selected ? 'selected' : '';
    const checked = this.props.selected ? true : false;
    const starred = this.props.starred ? 'fa-star' : 'fa-star-o';

    return (
      <div className={`row message ${read} ${selected}`}>
        <div className="col-xs-1">
          <div className="row">
            <div className="col-xs-2">
              <input type="checkbox" checked={checked} onChange={this.toggleSelected}/>
            </div>
            <div className="col-xs-2">
              <i className={`star fa ${starred}`} onClick={this.toggleStarred}></i>
            </div>
          </div>
        </div>
        <div className="col-xs-11">
          {this.props.labels.map((label) => {
            return <Label key={label} label={label} />
          })}
          <a href="/">
            {this.props.subject}
          </a>
        </div>
      </div>
    )
  }
}

export default Message;
