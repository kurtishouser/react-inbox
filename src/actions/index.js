export const MESSAGES_PENDING = 'MESSAGES_PENDING';
export const MESSAGES_RECEIVED = 'MESSAGES_RECEIVED';
export function fetchMessages() {
  return async (dispatch, getState, { Api }) => {

    if (getState().messages.ids.length) {
      return
    }

    dispatch({
      type: MESSAGES_PENDING
    });

    const json = await Api.fetchMessages();

    return dispatch({
      type: MESSAGES_RECEIVED,
      messages: json._embedded.messages
    });
  }
}

export const SEND_MESSAGE = 'SEND_MESSAGE';
export function sendMessage(message) {
  return async (dispatch, getState, { Api }) => {

    let body = message;

    const response = await Api.sendMessage(body);

    return dispatch({
      type: SEND_MESSAGE,
      response
    });
  }
}

export const FETCH_MESSAGE_BODY = 'FETCH_MESSAGE_BODY';
export function fetchMessageBody(id) {
  return async (dispatch, getState, { Api }) => {

  if (getState().messages.messagesById[id].body.length > 0) {
    // update read status just in case
    await Api.patchRequest({
      'messageIds': [id],
        'command': 'read',
        'read': true,
    });

    return;
  }

    const response = await Api.fetchMessageBody(id);

    // update read status
    await Api.patchRequest({
      'messageIds': [id],
        'command': 'read',
        'read': true,
    });

    return dispatch({
      type: FETCH_MESSAGE_BODY,
      id: response.id,
      messageBody: response.body
    })
  }
}

export const TOGGLE_SELECTED = 'TOGGLE_SELECTED';
export function toggleSelected(id) {
  return {
      type: TOGGLE_SELECTED,
      id
  }
}

export const TOGGLE_STARRED = 'TOGGLE_STARRED';
export function toggleStarred(id) {
  return async (dispatch, getState, { Api }) => {

    let body = {
        'messageIds': [id],
        'command': 'star',
        'star': !getState().messages.messagesById[id].starred,
    }

    const response = await Api.patchRequest(body);

    if (response === 200) {
      return dispatch({
        type: TOGGLE_STARRED,
        id
      });
    }
  }
}

export const TOGGLE_SELECT_ALL ='TOGGLE_SELECT_ALL';
export function toggleSelectAll(status) {
  return {
    type: TOGGLE_SELECT_ALL,
    status
  }
}

export const UPDATE_READ_STATUS = 'UPDATE_READ_STATUS';
export function updateReadStatus(status) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected
      // && getState().messages.messagesById[id].read !== status
    ));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
          'command': 'read',
          'read': status,
        }

      const response = await Api.patchRequest(body);

      if (response === 200) {
        return dispatch({
          type: UPDATE_READ_STATUS,
          status,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const ADD_LABEL = 'ADD_LABEL';
export function addLabel(label) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected &&
      !getState().messages.messagesById[id].labels.includes(label)));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'addLabel',
        'label': label,
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: ADD_LABEL,
          label,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const REMOVE_LABEL = 'REMOVE_LABEL';
export function removeLabel(label) {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected &&
      getState().messages.messagesById[id].labels.includes(label)));

    // only make the API call and dispatch action when there are IDs
    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'removeLabel',
        'label': label,
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: REMOVE_LABEL,
          label,
          messageIds // optimizes reducer
        });
      }
    }
  }
}

export const DELETE_MESSAGES= 'DELETE_MESSAGES';
export function deleteMessages() {
  return async (dispatch, getState, { Api }) => {

    let messageIds = getState().messages.ids.filter(id => (
      getState().messages.messagesById[id].selected));

    if (messageIds.length) {
      let body = {
        'messageIds': messageIds,
        'command': 'delete',
      }

      const response = await Api.patchRequest(body)

      if (response === 200) {
        return dispatch({
          type: DELETE_MESSAGES,
        });
      }
    }
  }
}
