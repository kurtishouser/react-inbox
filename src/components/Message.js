import React, {Component} from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';
import { toggleSelected, toggleStarred } from '../actions';
import { bindActionCreators } from 'redux';
import MessageBody from './MessageBody.js';

import PropTypes from 'prop-types';

export class Message extends Component {

  render() {
    const { id, selected, read, starred, subject, labels } = this.props.message;
    const isRead= read ? 'read' : 'unread';
    const isSelected = selected ? 'selected' : '';
    const isChecked = selected ? true : false;
    const isStarred = starred ? 'fa-star' : 'fa-star-o';
    const displayMessageBody = parseInt(this.props.match.params.id, 10) === id;

    return (
      <div>

        <div className={`row message ${isRead} ${isSelected}`}>
          <div className="col-xs-1">
            <div className="row">
              <div className="col-xs-2">
                <input type="checkbox" checked={isChecked} onChange={() => this.props.toggleSelected(this.props.messageId)}/>
              </div>
              <div className="col-xs-2">
                <i className={`star fa ${isStarred}`} onClick={() => this.props.toggleStarred(this.props.messageId)}></i>
              </div>
            </div>
          </div>
          <div className="col-xs-11">
            {labels.map((label) => {
              return <span key={label} className="label label-warning">{label}</span>
            })}
            <Link to={`/messages/${id}`}>{subject}</Link>
          </div>
        </div>
        {displayMessageBody &&
          <MessageBody messageId={id}/>
        }
    </div>
    )
  }
}

Message.propTypes = {
  message: PropTypes.object.isRequired
};

Message.defaultProps = {
  message: {
    labels: []
  },
  match: {params:{id:1}},
};

export const mapStateToProps = (state, ownProps) => {
  const message = state.messages.messagesById[ownProps.messageId];
  console.log(message);
  return {
    message,
  }
}

const mapDispatchToProps = dispatch => bindActionCreators({
  toggleSelected,
  toggleStarred
}, dispatch);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Message);
