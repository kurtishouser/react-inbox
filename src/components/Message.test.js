import React from 'react'
import { shallow, render, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Message, mapStateToProps } from './Message'
import { fetchMessages, PRODUCTS_RECEIVED } from '../actions'

describe('Message', () => {

  it('should render without any props defined', () => {
    const component = shallow(
      <Message match={{params:{id:1}}}/>
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should render with props passed in', () => {
    const component = shallow(
      <Message message={{
        _links: {
          self: {
            href: "http://localhost:8181/api/messages/1"
          }
        },
        id: 1,
        subject: "Hi",
        starred: true,
        read: true,
        labels: [
          "dev",
          "personal"
        ],
        body: "Hello there",
        selected: false,
      }} match={{params:{id:1}}} />
    );
    expect(toJson(component)).toMatchSnapshot();
  });

  it('mapStateToProps', () => {
    const state = {
      messages: {
        ids: [1],
        messagesById: {
          1: {
            _links: {
              self: {
                href: "http://localhost:8181/api/messages/1"
              }
            },
            id: 1,
            subject: "Hi",
            starred: true,
            read: true,
            labels: ["dev", "personal"],
            body: "Hello there",
            selected: false,
          }
        }
      }
    }
    const ownProps = { messageId: 1};
    const expected = {
      message: {
          _links: {
            self: {
              href: "http://localhost:8181/api/messages/1"
            }
          },
          id: 1,
          subject: "Hi",
          starred: true,
          read: true,
          labels: ["dev", "personal"],
          body: "Hello there",
          selected: false,
      }
    }
    expect(mapStateToProps(state, ownProps)).toEqual(expected)
  })

  xit('has a CHECKBOX that changes after click', () => {
    const toggleSelected = jest.fn();
    const message = shallow(
      <Message message={{
        _links: {
          self: {
            href: "http://localhost:8181/api/messages/1"
          }
        },
        id: 1,
        subject: "Hi",
        starred: true,
        read: true,
        labels: [
          "dev",
          "personal"
        ],
        body: "Hello there",
        selected: false,
      }} match={{params:{id:1}}} />,
    );

    const checkbox = message.find({ type: 'checkbox' });
    console.log('chk',checkbox);

    expect(checkbox.props().checked).toEqual(false);
    checkbox.simulate('change');
    expect(checkbox.props().checked).toEqual(true);
  });

  it('simulate star clicked', () => {
    const wrapper = shallow(
      <Message match={{params:{id:1}}}/>
    );
    console.log(wrapper);
    expect(wrapper).toMatchSnapshot();

    wrapper.find('.star.fa').simulate('click');

    expect(wrapper).toMatchSnapshot();
});

});
