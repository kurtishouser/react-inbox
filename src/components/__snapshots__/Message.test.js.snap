// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message should render with props passed in 1`] = `
<div>
  <div
    className="row message read "
  >
    <div
      className="col-xs-1"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-2"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
        <div
          className="col-xs-2"
        >
          <i
            className="star fa fa-star"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="col-xs-11"
    >
      <span
        className="label label-warning"
      >
        dev
      </span>
      <span
        className="label label-warning"
      >
        personal
      </span>
      <Link
        replace={false}
        to="/messages/1"
      >
        Hi
      </Link>
    </div>
  </div>
  <Connect(MessageBody)
    messageId={1}
  />
</div>
`;

exports[`Message should render without any props defined 1`] = `
<div>
  <div
    className="row message unread "
  >
    <div
      className="col-xs-1"
    >
      <div
        className="row"
      >
        <div
          className="col-xs-2"
        >
          <input
            checked={false}
            onChange={[Function]}
            type="checkbox"
          />
        </div>
        <div
          className="col-xs-2"
        >
          <i
            className="star fa fa-star-o"
            onClick={[Function]}
          />
        </div>
      </div>
    </div>
    <div
      className="col-xs-11"
    >
      <Link
        replace={false}
        to="/messages/undefined"
      />
    </div>
  </div>
</div>
`;

exports[`Message simulate star clicked 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <div
        className="row message unread "
    >
        <div
            className="col-xs-1"
        >
            <div
                className="row"
            >
                <div
                    className="col-xs-2"
                >
                    <input
                        checked={false}
                        onChange={[Function]}
                        type="checkbox"
                    />
                </div>
                <div
                    className="col-xs-2"
                >
                    <i
                        className="star fa fa-star-o"
                        onClick={[Function]}
                    />
                </div>
            </div>
        </div>
        <div
            className="col-xs-11"
        >
            <Link
                replace={false}
                to="/messages/undefined"
            />
        </div>
    </div>
</div>,
  "nodes": Array [
    <div>
      <div
            className="row message unread "
      >
            <div
                  className="col-xs-1"
            >
                  <div
                        className="row"
                  >
                        <div
                              className="col-xs-2"
                        >
                              <input
                                    checked={false}
                                    onChange={[Function]}
                                    type="checkbox"
                              />
                        </div>
                        <div
                              className="col-xs-2"
                        >
                              <i
                                    className="star fa fa-star-o"
                                    onClick={[Function]}
                              />
                        </div>
                  </div>
            </div>
            <div
                  className="col-xs-11"
            >
                  <Link
                        replace={false}
                        to="/messages/undefined"
                  />
            </div>
      </div>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <Message
        match={
                Object {
                        "params": Object {
                          "id": 1,
                        },
                      }
        }
        message={
                Object {
                        "labels": Array [],
                      }
        }
/>,
      "_debugID": 5,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": Message {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "match": Object {
            "params": Object {
              "id": 1,
            },
          },
          "message": Object {
            "labels": Array [],
          },
        },
        "refs": Object {},
        "state": null,
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 3,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <div
                    className="row message unread "
          >
                    <div
                              className="col-xs-1"
                    >
                              <div
                                        className="row"
                              >
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <input
                                                            checked={false}
                                                            onChange={[Function]}
                                                            type="checkbox"
                                                  />
                                        </div>
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <i
                                                            className="star fa fa-star-o"
                                                            onClick={[Function]}
                                                  />
                                        </div>
                              </div>
                    </div>
                    <div
                              className="col-xs-11"
                    >
                              <Link
                                        replace={false}
                                        to="/messages/undefined"
                              />
                    </div>
          </div>
</div>,
        "_debugID": 6,
        "_renderedOutput": <div>
          <div
                    className="row message unread "
          >
                    <div
                              className="col-xs-1"
                    >
                              <div
                                        className="row"
                              >
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <input
                                                            checked={false}
                                                            onChange={[Function]}
                                                            type="checkbox"
                                                  />
                                        </div>
                                        <div
                                                  className="col-xs-2"
                                        >
                                                  <i
                                                            className="star fa fa-star-o"
                                                            onClick={[Function]}
                                                  />
                                        </div>
                              </div>
                    </div>
                    <div
                              className="col-xs-11"
                    >
                              <Link
                                        replace={false}
                                        to="/messages/undefined"
                              />
                    </div>
          </div>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Message
    match={
        Object {
            "params": Object {
              "id": 1,
            },
          }
    }
    message={
        Object {
            "labels": Array [],
          }
    }
/>,
}
`;
